libarchive/archive_write_add_filter_xz.c

```diff
@@ -245,27 +245,40 @@ archive_compressor_xz_init_stream(struct archive_write_filter *f,
 	} else if (f->code == ARCHIVE_FILTER_LZMA) {
 		ret = lzma_alone_encoder(&(data->stream), &data->lzma_opt);
 	} else {	/* ARCHIVE_FILTER_LZIP */
-		int dict_size = data->lzma_opt.dict_size;
-		int ds, log2dic, wedges;
+		uint32_t dict_size = data->lzma_opt.dict_size;
+		uint32_t wedges;
+		unsigned char ds;
 
 		/* Calculate a coded dictionary size */
-		if (dict_size < (1 << 12) || dict_size > (1 << 29)) {
+		if (dict_size < (1U << 12) || dict_size > (1U << 29)) {
 			archive_set_error(f->archive, ARCHIVE_ERRNO_MISC,
 			    "Unacceptable dictionary size for lzip: %d",
 			    dict_size);
 			return (ARCHIVE_FATAL);
 		}
-		for (log2dic = 29; log2dic >= 12; log2dic--) {
-			if (dict_size & (1 << log2dic))
-				break;
+
+		{
+#if __has_builtin(__builtin_clz)
```

Comment:
You have to protect uses of `__has_builtin` appropriately.  See the example at https://gcc.gnu.org/onlinedocs/cpp/_005f_005fhas_005fbuiltin.html for how to correctly use this.